working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+  final List<Object> v;
+}
+
+C1<Shareable>(); // ok
+C2<Shareable>(); // runtime error: static type of C2.v is not shareable
+```
+
+A simpler alternative would be to introduce a set of shareable collection types: `ShareableList<T>`, `ShareableSet<T>`,  `ShareableMap<K, V>`, all with appropriate type parameter bounds restricting elements to subtypes of `Shareable?`. Builtin implementation of `List<T>` will also implement `ShareableList<T>` when `T` is shareable, but there would be no implicit expectation of shareability built into `List<T>` itself.
+
+### Shareable core types
+
+The following types will become shareable:
+
+* `num` (`int` and `double`), `String`, `bool`, `Null`
+* `Enum` - meaning that all enums will be shareable, this is fine because enums are deeply immutable. 
+* `RegExp`, `DateTime`, `Uri` - these might require work to be shareable, but it makes sense to allow sharing them.
+* `TypedData` - which makes all typed data types shareable.
+  * **BREAKING CHANGE**: making `TypedData` shareable changes the behaviour of `SendPort` which will stop copying it. It's unclear if we want to maintain old behaviour for compatibility reasons. One option here is to say that `TypedData` is only shared when sending through `SendPort` iff it is a member of another explicitly `Shareable` type.    
+* `Pointer` 
+
+## Shared Isolates
+
+Lets take another look at the following example: 
+
+```dart
+int global = 0; 
+
+void main() async {
+  global = 42;
+  await Isolate.run(() {
+    print(global);  // => 0
+    global = 24;
+  });  
+  print(global);  // => 42
+}
+```
+
+Stripped to the bare minimum the example does not seem to behave in a confusing way: it seems obvious that each isolate has its own version of `global` variable and mutations of `global` are not visible across isolates. However, in the real world code such behavior might be hidden deep inside a third party dependency and thus much harder to detect and understand. This behavior also makes interoperability with native code more awkward than it ought to be: calling Dart requires an isolate, something that native code does not really know or care about. 
+
+```dart
+int global;
+
+@pragma('vm:entry-point')
+void foo() => global++;
+```
+
+`@shared` global variables allow developers to tackle this problem - but hidden dependency on global state might introduce hard to diagnose and debug bugs. 
+
+We propose to tackle this problem by introducing the concept of _shared isolate_: **code running in a _shared isolate_ can only access `@shared` state and not any of isolated state, an attempt to access isolated state results in a dynamic `IsolationError`**.
+
+``` dart
+int global = 0; 
+
+@shared
+int sharedGlobal = 0;
+
+void main() async {
+  global = 42;
+  sharedGlobal = 42;
+  await Isolate.run(() {
+    print(global);  // IsolationError: Can't access 'global' when running in shared isolate
+    global = 24;  // IsolationError: Can't access 'global' when running in shared isolate
```

Comment:
It would be useful to have some way to pass a reference to non-shareable object to another thread as an opaque value. No thread except the owner thread can dereference it, but it's OK to pass it around threads as long as non-owner threads do not look inside.

For example, `dart:concurrent` could include a special type `Box<T>` where `T` allows any Dart type, including non-shared references. The API could be something along the following lines:

```dart
/// Stores a value of type `T` that can be passed to
/// arbitrary threads, but only accessed by the owner
/// thread.
final class Box<T> implements Shareable {
  /// Creates a new box that stores a value.
  ///
  /// The thread that called this constructor is the owner
  /// of this box and is the only thread that can call [read].
  Box(T value);

  /// Returns the value that was passed to the [Box] constructor.
  ///
  /// This method may only be called by the thread that owns
  /// this box. Throws [IsolationError] if called from a thread
  /// that does not own this box. 
  T read();
}
```

Example usage:

```dart
class Foo {} // a non-shareable class!

void main() async {
  final foo = Foo();
  final boxedFoo = Box<Foo>(foo);
  await Isolate.run(() {
    final boxedFoos = <Box<Foo>>[];
    boxedFoos.add(boxedFoo);  // OK!

    boxedFoo.read(); // IsolationError

    final foos = <Foo>[];
    foos.add(foo); // IsolationError
  });

  boxedFoo.read(); // OK!
}
```